### PR TITLE
Bridge input

### LINKED ISSUE #18591: Update UI and data model (re-frame) on Select Input Page to handle Bridging flow
The transaction confirmation page needs a few adjustments to handle the bridging
designs
https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=1344-291019&mode=design&t=RuyhufmiW7Y5H6pU-4

Also ensure the re-frame data model is correctly setup and that the correct data is passed to the next step of the flow.

This task probably depends on some other issues being completed in advance so check the associated epic 👍

refactor where needed so that the screens handle the various transaction types - i.e bridigin, send-asset, send collectible

### PR DESCRIPTION
**Pull request by Pau1fitz**
Closes #18591

Figma design [here](https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=1344-291019&mode=design&t=Ebjm9JvrEgDfXynD-0).

https://github.com/status-im/status-mobile/assets/6873736/9e1177fe-223f-46a6-9116-f0fdfa325982

iOS
Android

Testing notes:
Navigate to an account screen
click on Bridge
click on network
enter amount

### CODE REVIEW

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (OmarBasem) [inline src/status_im/contexts/wallet/send/input_amount/view.cljs]: I think better keep crypto decimals calculation from `(utils/get-crypto-decimals-count token)` here because now it is being calculated in two places

File: src/status_im/contexts/wallet/events.cljs
Comment (briansztamfater) [inline src/status_im/contexts/wallet/events.cljs]: We should dispatch `:navigate-to-within-stack` event here

File: src/status_im/contexts/wallet/events.cljs
Comment (briansztamfater) [inline src/status_im/contexts/wallet/events.cljs]: We should dispatch `:navigate-to-within-stack` event here

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (briansztamfater) [inline src/status_im/contexts/wallet/send/input_amount/view.cljs]: I think having these type of checks is not scalable (this one and those in `:button-one-label` and `:button-one-props`), maybe we would end having lots of these ones all over the file when we start to add more functionality or if we need this screen to support another transaction type. Maybe a good idea would be to abstract this into a common component, and then define two different screens with different configurable parameters, similar to what we did with `status-im.common.scan-qr-code.view` namespace, but inside wallet namespace, maybe in `status-im.contexts.wallet.common`. What do you think?

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (Pau1fitz) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs]: good shout @OmarBasem. The reason for this was that the `total-amount` was not originally included on the token. I have now updated this so we no longer calculate twice and `total-value` is already calculated for the token when the `input-amount` component receives it.

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (Pau1fitz) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs]: thanks for the feedback @briansztamfater, makes sense. Just to confirm do you mean to extend the props that `input-amount` can take. i.e. on-navigate-back, button label etc.? That is what I understand from the `status-im.common.scan-qr-code.view` but maybe am missing something.

File: src/status_im/contexts/wallet/account/bridge_send/style.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/account/bridge_send/style.cljs]: does `flex 1` work here instead? 🤔 
I think we generally avoid percentages on view widths/heights in the codebase

File: src/status_im/contexts/wallet/account/bridge_send/view.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/account/bridge_send/view.cljs]: this can be `[quo.theme]` as we don't need to alias it.

File: src/status_im/contexts/wallet/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/events.cljs]: e.g 
` :fx [[:dispatch [:navigate-to-within-stack [:wallet-bridge-to stack-id]]]]`

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/input_amount/view.cljs]: if possible we use a different name than `type` here because it is actually a cljs function.
`transfer-type` is better? 🤔

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/input_amount/view.cljs]: perhaps `on-navigate-back` should be passed in as a prop from the wrapper page?
cc @briansztamfater wdyt?

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/input_amount/view.cljs:289]: @briansztamfater, what do you think is the cleanest way to pass along the bottom actions? 
Perhaps we just structure it so that the wrapper page has the bottom-actions rendered there?

e.g as children
```
(defn- view-internal
  []
  (let [send-bridge-data (rf/sub [:wallet/wallet-send])
        send-type        (:type send-bridge-data)]
    [rn/view {:style style/bridge-send-wrapper}
     [input-amount/view
      {:type send-type}
      [quo/bottom-actions ...]]
      ]))
 ```     
 or a view outside the input-amount

```
(defn- view-internal
  []
  (let [send-bridge-data (rf/sub [:wallet/wallet-send])
        send-type        (:type send-bridge-data)]
    [rn/view {:style style/bridge-send-wrapper}
     [input-amount/view
      {:type send-type}]
      [quo/bottom-actions ...]
      ]))
 ```     
 thoughts ?

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (briansztamfater) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs]: Yep I think we should pass on-navigate-back as a prop

Comment (Pau1fitz) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs]: have updated to pass `navigate-back` as a prop. Is it also worth adding the button label and props as a prop or do we want to take a different approach

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (OmarBasem) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs]: @Pau1fitz I still see it being called in two places, in `send_amount/view` and in `bridge_send/view`

> The reason for this was that the total-amount was not originally included

Which total-amount or total-value are you referring to?

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (briansztamfater) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs:289]: Not sure what you mean @J-Son89, I think you posted the same example twice 🤔 . But I think it depends on how customizable we want it this component to be, I think if its just the buttons texts we want to customize we can have some simple props in the wrapper components.

Comment (J-Son89) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs:289]: Sounds good!

File: src/status_im/contexts/wallet/send/input_amount/view.cljs
Comment (Pau1fitz) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs]: @OmarBasem not sure I'm following. Are you looking at the most up to date version?

**Review by OmarBasem — APPROVED** [review]

**Review by briansztamfater — APPROVED** [review]

File: src/status_im/contexts/wallet/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/events.cljs]: Fix this `->` code style.

the como style is
```clojure
(-> db
    f
    g
    h)
```

**Review by ulisesmac — APPROVED** [review]

**Comment by J-Son89** [comment]
@status-im/mobile-qa - Can we get some regression testing on this pr?
This feature Bridgin is blocked by a feature flag, however it shares some code with the sending flows. In that case I think it's safer we check that a transaction can be made here.

**Comment by pavloburykh** [comment]
> @status-im/mobile-qa - Can we get some regression testing on this pr? This feature Bridgin is blocked by a feature flag, however it shares some code with the sending flows. In that case I think it's safer we check that a transaction can be made here.

Hi @J-Son89! I hope tomorrow someone will be able to take it but there might be delay, sorry for that. Currently part of the team is busy with testing Communities functionality which is needed for upcoming release.

**Comment by briansztamfater** [comment]
Just thinking that `:wallet/get-suggested-routes` event is prepared only for transfer transactions, not bridging, and must be adapted in order to make bridging actually work, so routes shown on this PR wouldn't be accurate (can't bridge a token from-to the same network as the video in the description shows with Optimism). What do you think @J-Son89, should we create a separate issue for that, or handle it on this one?
Also asking to avoid unnecessary testing efforts from QA while testing this PR.

**Comment by J-Son89** [comment]
thanks @pavloburykh, that's fine! I'll look into what @briansztamfater is suggesting here in the interim anyway 👍 
So we can hold off testing this until we've checked that 👌

**Comment by pavloburykh** [comment]
> thanks @pavloburykh, that's fine! I'll look into what @briansztamfater is suggesting here in the interim anyway 👍 So we can hold off testing this until we've checked that 👌

Cool. Ping @VolodLytvynenko when PR will be ready for QA.

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/events.cljs:121]: @briansztamfater - can you take a look at this again now?
this function is updated now to set `disabled-to-chain-ids` by taking all possible networks and removing the selected `bridge-to-chain-id`. Also sets the transaction type param using `(get-in db [:wallet :ui :send :type])` - this is also added for collectibles. For now it defaults to send asset but we can add this property to the other send asset flows too 👍

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/events.cljs:119]: this was just updated for readability
`[:wallet/get-suggested-routes 1]` vs `[:wallet/get-suggested-routes {:amount 1}]` in the second it's more apparent what 1 relates to.

File: src/status_im/contexts/wallet/send/events.cljs
Comment (briansztamfater) [inline src/status_im/contexts/wallet/send/events.cljs:148]: In general I think the approach is fine, but...

I think transaction types should be just `:send` or `:bridge`. We could potentially bridge collectibles in the future, so the code should be flexible enough to support this. A collectible is just a different kind of token, so I am not convinced to have a transaction type of `:collectible`, that could lead to confusion in the future.

If we only have `:send` and `:bridge` transaction types, then checking if any of`[:wallet :ui :send :collectible]` or `[:wallet :ui :send :token]` are defined in app-db would be enough for setting the correct `transaction-type-param` required by status-go endpoint.

cc @J-Son89 @Pau1fitz

Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs:148]: true but `:bridge` is just a form of sending as well? 
Mostly `transfer-type` here is related to the parameter of the `get-suggested-routes` endpoint. We know in advance depending on the flow of which type we want so why should we leave it up to this function to resolve or figure out which type of transaction it is.
The parameters of this function some what depend on the different send flows too.

If in the future we are bridging a collectible the transaction type would be directly related to the value needed for `get-suggested-routes` endpoint.
I'm happy either way just giving the thought behind this 👍

Comment (briansztamfater) [reply src/status_im/contexts/wallet/send/events.cljs:148]: Also, for another issue, but we may consider refactor the db path where we store transaction data from `:wallet :ui :send` to `:wallet :ui :transaction` (or `:wallet :ui :tx` if we want shorter names)

Comment (briansztamfater) [reply src/status_im/contexts/wallet/send/events.cljs:148]: I think of bridging as something different of sending.

- Send: is a wallet to wallet transaction
- Bridge: Process of moving user tokens from one network to other network(s) through a bridge

Sending may involve bridging (a batch of multiple and different kind transactions) if the response of getSuggestedRoutes involves a bridging transaction, but at this point in the flow we should just know if it is a send or bridge, and which token we want to send or bridge.

Anyways, as I said, I think that the approach is fine for now, I may propose changes in the future and we can discuss them later on. In the end, what I don't really like is how status-go defines the transaction types, and I would like to abstract mobile logic from it instead of matching 1-1. I am thinking also of changing the path in app-db to be `[:wallet :ui :transaction]` instead of `[:wallet :ui :send]` to make it more flexible, but I need to think it better while we integrate both flows and that may require a bigger refactor in the case we agree to do it.

**Comment by J-Son89** [comment]
@VolodLytvynenko - can we check this pr whenever you have a chance? It just needs a regression test particularly with sending tokens and collectibles.

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs:148]: Sounds good @briansztamfater, happy to go with whatever you suggest here 👍

**Comment by VolodLytvynenko** [comment]
hi @Pau1fitz Thank you for PR. No issues from my side. PR is ready to be merged

**Pull request merged**
